proxy/src/control/cache.rs

```diff
@@ -0,0 +1,199 @@
+use std;
+use std::collections::HashSet;
+use std::mem;
+
+/// A cache that supports incremental updates with lazy resetting on
+/// invalidation.
+///
+/// When the cache `c` initially becomes invalid (i.e. it becomes
+/// potentially out of sync with the data source so that incremental updates
+/// would stop working), call `c.reset_on_next_modification()`; the next
+/// incremental update will then replace the entire contents of the cache,
+/// instead of incrementally augmenting it. Until that next modification,
+/// however, the stale contents of the cache will be made available.
+pub struct Cache<T> {
+    values: HashSet<T>,
```

Comment:
I wonder if we might expect some performance gains from switching to `IndexSet`...

Reply:
Let's make any functional changes (beyond moving the code) in a separate PR.